### PR TITLE
Add Huber loss

### PR DESCRIPTION
**Pull request by henry0312**
Add [Huber loss](https://en.wikipedia.org/wiki/Huber_loss).
Thanks!

### CODE REVIEW

**Comment by henry0312** [comment]
I'm making objective function now.

**Comment by henry0312** [comment]
@guolinke ready for review

File: include/LightGBM/config.h [hunk 1]
```diff
@@ -137,6 +137,8 @@ struct ObjectiveConfig: public ConfigBase {
 public:
   virtual ~ObjectiveConfig() {}
   double sigmoid = 1.0f;
+  // for Huber loss
+  double delta = 1.0f;
```
Comment (wxchan) [inline include/LightGBM/config.h]: better rename to `huber_delta`

File: include/LightGBM/config.h [hunk 2]
```diff
@@ -196,6 +200,7 @@ struct BoostingConfig: public ConfigBase {
 public:
   virtual ~BoostingConfig() {}
   double sigmoid = 1.0f;
+  double delta = 1.0f;
```
Comment (wxchan) [inline include/LightGBM/config.h]: BoostingConfig needs delta?

File: python-package/lightgbm/sklearn.py [hunk 3]
```diff
@@ -490,6 +493,31 @@ def feature_importance(self):
...
+                                            n_estimators=n_estimators, max_bin=max_bin,
+                                            subsample_for_bin=subsample_for_bin, objective=objective,
+                                            min_split_gain=min_split_gain, min_child_weight=min_child_weight,
+                                            min_child_samples=min_child_samples, subsample=subsample,
+                                            subsample_freq=subsample_freq, colsample_bytree=colsample_bytree,
+                                            reg_alpha=reg_alpha, reg_lambda=reg_lambda,
+                                            scale_pos_weight=scale_pos_weight, is_unbalance=is_unbalance,
+                                            seed=seed, nthread=nthread, silent=silent,
+                                            sigmoid=sigmoid, huber_delta=huber_delta,
+                                            max_position=max_position, label_gain=label_gain,
```
Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: remove scale_pos_weight, is_unbalance, sigmoid, max_position and label_gain, they are not available for regression task.

File: src/metric/regression_metric.hpp [hunk 4]
```diff
@@ -110,13 +113,34 @@ class L1Metric: public RegressionMetric<L1Metric> {
...
+  inline static score_t LossOnPoint(float label, score_t score, float delta=0.0f) {
     return std::fabs(score - label);
   }
   inline static const char* Name() {
     return "l1";
   }
 };
 
+/*! \brief Huber loss for regression task */
+class HuberLoss: public RegressionMetric<HuberLoss> {
```
Comment (wxchan) [inline src/metric/regression_metric.hpp]: better rename to `HuberLossMetric`

**Comment by henry0312** [comment]
@wxchan thank you for your review. updated.

File: docs/Parameters.md [hunk 5]
```diff
@@ -16,8 +16,9 @@ The parameter format is ```key1=value1 key2=value2 ... ``` . And parameters can
 * ```task```, default=```train```, type=enum, options=```train```,```prediction```
   * ```train``` for training
   * ```prediction``` for prediction.
-* ```application```, default=```regression```, type=enum, options=```regression```,```binary```,```lambdarank```,```multiclass```, alias=```objective```,```app```
+* ```application```, default=```regression```, type=enum, options=```regression```,```huber_loss```,```binary```,```lambdarank```,```multiclass```, alias=```objective```,```app```
   * ```regression```, regression application
+  * ```huber_loss```, regression application
```
Comment (guolinke) [inline docs/Parameters.md]: more details here? or a link

Comment (guolinke) [inline docs/Parameters.md]: change name to ```huber``` ? keep it simple is better

File: src/metric/metric.cpp [hunk 6]
```diff
@@ -11,6 +11,8 @@ Metric* Metric::CreateMetric(const std::string& type, const MetricConfig& config
     return new L2Metric(config);
   } else if (type == std::string("l1")) {
     return new L1Metric(config);
+  } else if (type == std::string("huber_loss")) {
```
Comment (guolinke) [inline src/metric/metric.cpp]: change name to ```huber```?

File: src/metric/regression_metric.hpp [hunk 7]
```diff
@@ -110,13 +113,34 @@ class L1Metric: public RegressionMetric<L1Metric> {
...
+        const double diff = score - label;
+        if (std::abs(diff) <= delta) {
+            return 0.5 * diff * diff;
+        } else {
+            return delta * (std::abs(diff) - 0.5 * delta);
+        }
+    }
+
+    inline static const char* Name() {
+        return "huber_loss";
```
Comment (guolinke) [inline src/metric/regression_metric.hpp]: change name to ```huber```

File: src/metric/regression_metric.hpp [hunk 8]
```diff
@@ -110,13 +113,34 @@ class L1Metric: public RegressionMetric<L1Metric> {
 public:
   explicit L1Metric(const MetricConfig& config) :RegressionMetric<L1Metric>(config) {}
 
-  inline static score_t LossOnPoint(float label, score_t score) {
+  inline static score_t LossOnPoint(float label, score_t score, float delta=0.0f) {
```
Comment (guolinke) [inline src/metric/regression_metric.hpp]: to avoid warning, changed to ```(float label, score_t score, float)```.

File: src/metric/regression_metric.hpp [hunk 9]
```diff
@@ -91,7 +94,7 @@ class L2Metric: public RegressionMetric<L2Metric> {
 public:
   explicit L2Metric(const MetricConfig& config) :RegressionMetric<L2Metric>(config) {}
 
-  inline static score_t LossOnPoint(float label, score_t score) {
+  inline static score_t LossOnPoint(float label, score_t score, float delta=0.0f) {
```
Comment (guolinke) [inline src/metric/regression_metric.hpp]: to avoid warning, changed to ```(float label, score_t score, float)```.

File: src/metric/regression_metric.hpp [hunk 10]
```diff
@@ -73,6 +72,10 @@ class RegressionMetric: public Metric {
     return sum_loss / sum_weights;
   }
 
+protected:
+  /*! \brief delta for Huber loss */
+  double huber_delta_;
```
Comment (guolinke) [inline src/metric/regression_metric.hpp]: give a initial value (e.g. 1.0f) in the constructor of ```RegressionMetric```

File: src/metric/regression_metric.hpp [hunk 11]
```diff
@@ -110,13 +113,34 @@ class L1Metric: public RegressionMetric<L1Metric> {
...
 };
 
+/*! \brief Huber loss for regression task */
+class HuberLossMetric: public RegressionMetric<HuberLossMetric> {
+public:
+    explicit HuberLossMetric(const MetricConfig& config) :RegressionMetric<HuberLossMetric>(config) {
+        huber_delta_ = config.huber_delta;
+    }
+
+    inline static score_t LossOnPoint(float label, score_t score, float delta=1.0f) {
```
Comment (guolinke) [inline src/metric/regression_metric.hpp]: not default value for ```delta``` here

File: src/objective/objective_function.cpp [hunk 12]
```diff
@@ -9,6 +9,8 @@ namespace LightGBM {
 ObjectiveFunction* ObjectiveFunction::CreateObjectiveFunction(const std::string& type, const ObjectiveConfig& config) {
   if (type == std::string("regression")) {
     return new RegressionL2loss(config);
+  } else if (type == std::string("huber_loss")) {
```
Comment (guolinke) [inline src/objective/objective_function.cpp]: change name to ```huber```

File: src/objective/regression_objective.hpp [hunk 13]
```diff
@@ -51,5 +51,72 @@ class RegressionL2loss: public ObjectiveFunction {
...
+              gradients[i] = -delta_ * weights_[i];
+            }
+            hessians[i] = 0.0;
+        }
+      }
+    }
+  }
+
+  const char* GetName() const override {
+    return "huber_loss";
```
Comment (guolinke) [inline src/objective/regression_objective.hpp]: change name to ```huber```

**Comment by henry0312** [comment]
@guolinke updated :)
